docs/specs/lcb-signals-and-events.md

```diff
@@ -0,0 +1,222 @@
+
+1. Signal emission and handling is not very well checked.  At a language level, there is no way for widgets to describe what type of signals they can emit.  Similarly, there is no syntax for distinguishing between widget handlers and "normal" handlers, meaning that (in principal) a handler that is not intended to be a signal handler could end up getting called in a signal context if a child widget happens to start emitting a signal with a matching name.  Finally, there is no facility for checking that signal prototypes match signal handler prototypes.
+
+2. Similar issues to 1. above apply for event handling.
+
+3. In LCS, event handling propagates upward: if a button doesn't have an `MouseUp` handler, then the event gets passed on to a parent card, and so on.  Furthermore, there is the idea of explicit propagation: it is possible to do something in an event handler and then, for example, `pass MouseUp` to invoke the normal event propagation logic.
+
+   In LCB, event propagation is *not* currently automatic.  If a widget does not explicitly handle and re-`post` each possible event, then its script (or parent widget) never sees the event.
+
+4. Sometimes, it's important to be able to be able to prevent child widgets from seeing an event, or performing some pre-processing / post-processing around the normal event handling process.  The current LCB event handling scheme does not permit this.
+
+## Assumptions
+
+For the purposes of this document we assume that:
+
+* All connection of signal and event handlers is implicit, rather than explicit.
+
+* An instance of a widget as an engine control can be considered as a composed widget, comprising a script and the "real" widget instance (this makes sense when you consider the fact that the widget control exists, has a script, and can be manipulated even if the "real" widget was unable to be loaded for some reason).  This will be referred to as a "host widget" in the rest of this document.
+
+## Signals
+
+### Signal emission
+
+A widget's LCB source code must explicitly name, and provide prototypes for, the signals that it can emit [1]:
+
+    public signal <name:id>(<params:formals>)
+
+Signal declarations must be placed in module scope.  The signal **name** must be unique in the module.
+
+Signals are explicitly emitted by a widget using an `emit` statement:
+
+    emit <name:id>(<params:expressions>)
+
+It is a run-time error to emit a signal with the wrong number or types of emission parameters [2].  The compiler may statically verify that signals are emitted with compatible parameter lists.
+
+The `emit` statement may return before all the signal handlers for the signal have finished running [3].
+
+The signal handlers for a signal get run exactly once for each emission.  All of the handlers for an emission must complete before any of the handlers for a subsequent emission begin [4].
+
+The signal handlers for a signal may be run in any order.
+
```

Comment:
@trevordevore said:
>Another example is a document object that has a visual representation. As changes are made to the document you don't want to have to update the data in the dataview and update the data in the actual model.
Realistically any document view will require cached data to make things render efficiently - there is no difference between that and a dataview. I don't see the problem of a dataview control caching formatted data for it to display - as long as the cache can be trimmed as necessary if more memory is needed for the main app.

So you aren't updating the data in two places - you are storing an 'abstract' set of data in the app which is then used by the view controls to display the data... Caching of formatted data in such view controls is perfectly reasonable.

For example, the field's abstract data is an attributed string but it requires a whole lot of concrete (layout related) data to actually render. Although not quite structured like that at the moment, notionally the field control turns the abstract attributed string into layout structures as needed, dumping appropriate parts of the cached layout data when the field content changes.

You could imagine it structured as this... First you have an inner 'DataViewDisplay' which requires data to be provided to it, ready for display via a callback; then wrapping this you could have the actual 'DataView' wrapping it which provides the caching of formatted data requested async from the host. I'd probably suggest that (morally) the DataViewDisplay would be a semi-concrete baseclass with a overridable method for fetching the data. (By the way, I'm talking conceptually here - at the moment you'd have to write the DataView as a combined DataView/DataViewDisplay).